This notebook will help you learn about 3 main packages used for data analysis in Python, <a
href='https://pandas.pydata.org/'>pandas</a>, <a href='https://www.scipy.org/'>scipy</a> and <a href='https://numpy.org/'>
numpy </a>. pandas is the most popular Python library for data analysis and manipulation whereas numpy and scipy are
packages for scientific computing.

To use these packages in your notebook, import them as shown below:

In [2]:
import pandas as pd
import numpy as np
import scipy

To understand the functionality of these packages we will use the sample dataset in the "Data" folder.

## Reading files using pandas

The pandas <a href='https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html'>User Guide</a> is a useful and highly
recommended resource. Please refer to that for additional details on the commands used in this notebook.

First, the sample data can be read into a table (dataframe) using the read_csv command from pandas.
The documentation for this command can be found <a href="https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-read-csv-table">here</a>.

In [3]:
guide_annotations = pd.read_csv('../Data/sample_dataset_gene_annotations.csv')

To view the table that was just read, we use the head() command. It displays the top 5 rows of the table by default. 

In [4]:
guide_annotations.head()

sgRNA Sequence Annotated Gene Symbol Annotated Gene ID
0  AAAAAAAATCCGGACAATGG              SLC25A24             29957
1  AAAAAAAGGATGGTGATCAA               FASTKD3             79072
2  AAAAAAATGACATTACTGCA                 BCAS2             10286
3  AAAAAAATGTCAGTCGAGTG                 GPR18              2841
4  AAAAAACACAAGCAAGACCG                ZNF470            388566

We can then view the size of the read dataframe using the <i>shape</i> attribute.

In [5]:
guide_annotations.shape

(77441, 3)

This indicates that there are 77,441 rows and 3 columns in this table. 

## Indexing, Slicing

The following code blocks demonstrate subsetting rows and columns that satisfy a particular condition.

In [11]:
#To retrieve the rows for SLC25A24
gene_rows = guide_annotations[guide_annotations['Annotated Gene Symbol'] == 'SLC25A24']

gene_rows

sgRNA Sequence Annotated Gene Symbol Annotated Gene ID
0      AAAAAAAATCCGGACAATGG              SLC25A24             29957
27139  CCAGGCACTGGACCGCAATG              SLC25A24             29957
29333  CCGTCTGAAAATCATGATGC              SLC25A24             29957
54696  GGTATCCGCTCGCTTTGGAG              SLC25A24             29957

In [12]:
#Retrieve only the guides sequences for SLC25A24
guide_seqs = gene_rows['sgRNA Sequence']

guide_seqs

0        AAAAAAAATCCGGACAATGG
27139    CCAGGCACTGGACCGCAATG
29333    CCGTCTGAAAATCATGATGC
54696    GGTATCCGCTCGCTTTGGAG
Name: sgRNA Sequence, dtype: object

In [19]:
#Subsetting rows and getting a specific column in a single command
#Label-based selection
guides = guide_annotations.loc[guide_annotations['Annotated Gene Symbol'] == 'SLC25A24','sgRNA Sequence']

guides

0        AAAAAAAATCCGGACAATGG
27139    CCAGGCACTGGACCGCAATG
29333    CCGTCTGAAAATCATGATGC
54696    GGTATCCGCTCGCTTTGGAG
Name: sgRNA Sequence, dtype: object

In [18]:
#Subsetting rows and getting a specific column
#Index-based selection
#Get first 20 guide sequences
first_20_guides = guide_annotations.iloc[0:20,0]

first_20_guides

0     AAAAAAAATCCGGACAATGG
1     AAAAAAAGGATGGTGATCAA
2     AAAAAAATGACATTACTGCA
3     AAAAAAATGTCAGTCGAGTG
4     AAAAAACACAAGCAAGACCG
5     AAAAAACAGATGCCACCTGT
6     AAAAAACCAAACTTTGAAGT
7     AAAAAACCCGTAGATAGCCT
8     AAAAAACCTGGGCAAAACAG
9     AAAAAACCTTCAAAGTACAA
10    AAAAAACGAACGGATTAACA
11    AAAAAACTCTACCACCCAGT
12    AAAAAACTGGAGTACGACAG
13    AAAAAACTTACAAGTCCAGG
14    AAAAAACTTATGTCCATGGA
15    AAAAAACTTGTATCAGCTGT
16    AAAAAAGAATGCCCCAAAGG
17    AAAAAAGACCCCATCCGAGA
18    AAAAAAGCATTGCCTGACAG
19    AAAAAAGCCCTCACCAACTG
Name: sgRNA Sequence, dtype: object

In [20]:
# Subset columns
subset_columns = guide_annotations[['sgRNA Sequence','Annotated Gene Symbol']]

subset_columns

sgRNA Sequence Annotated Gene Symbol
0      AAAAAAAATCCGGACAATGG              SLC25A24
1      AAAAAAAGGATGGTGATCAA               FASTKD3
2      AAAAAAATGACATTACTGCA                 BCAS2
3      AAAAAAATGTCAGTCGAGTG                 GPR18
4      AAAAAACACAAGCAAGACCG                ZNF470
...                     ...                   ...
77436  TTTGTTTGGATTGCTTCAAG                 NEPRO
77437  TTTTACCTTGTTCACATGGA        NO_CURRENT_997
77438  TTTTGACTCTAATCACCGGT        NO_CURRENT_998
77439  TTTTTAATACAAGGTAATCT        NO_CURRENT_999
77440  TTTTTCTCACCCGATGAATC       NO_CURRENT_1000

[77441 rows x 2 columns]

## Merging

We often handle files that need to be combined at some stage of the analysis pipeline. The merge command in pandas is very useful for this purpose. 

Here, we will first read the tab with the read counts from Supplementary Data 1 and then merge it with the gene annotations we have. 